shell/platform/embedder/embedder.cc

```diff
@@ -0,0 +1,220 @@
+    weak_engine = platform_view->engine().GetWeakPtr(),  //
+    assets = std::move(assets),                          //
+    main = std::move(main),                              //
+    packages = std::move(packages)                       //
+  ] {
+    if (auto engine = weak_engine) {
+      engine->RunBundleAndSource(assets, main, packages);
+    }
+  });
+
+  *engine_out = reinterpret_cast<FlutterEngine>(
+      new PlatformViewHolder(std::move(platform_view)));
+
+  return kSuccess;
+}
+
+FlutterResult FlutterEngineShutdown(FlutterEngine engine) {
+  if (engine == nullptr) {
+    return kInvalidArguments;
+  }
+  // TODO(chinmaygarde): This is currently unsupported since none of the mobile
+  // shells need to do this. Add support and patch this call.
+  delete reinterpret_cast<PlatformViewHolder*>(engine);
+  return kSuccess;
+}
+
+FlutterResult FlutterEngineSendWindowMetricsEvent(
+    FlutterEngine engine,
+    const FlutterWindowMetricsEvent* flutter_metrics) {
+  if (engine == nullptr || flutter_metrics == nullptr) {
+    return kInvalidArguments;
+  }
+
+  auto holder = reinterpret_cast<PlatformViewHolder*>(engine);
+
+  blink::ViewportMetrics metrics;
+
+  metrics.physical_width = SAFE_ACCESS(flutter_metrics, width, 0.0);
+  metrics.physical_height = SAFE_ACCESS(flutter_metrics, height, 0.0);
+  metrics.device_pixel_ratio = SAFE_ACCESS(flutter_metrics, pixel_ratio, 1.0);
+
+  blink::Threads::UI()->PostTask(
+      [ weak_engine = holder->view()->engine().GetWeakPtr(), metrics ] {
+        if (auto engine = weak_engine) {
+          engine->SetViewportMetrics(metrics);
+        }
+      });
+  return kSuccess;
+}
+
+inline blink::PointerData::Change ToPointerDataChange(
+    FlutterPointerPhase phase) {
+  switch (phase) {
+    case kCancel:
+      return blink::PointerData::Change::kCancel;
+    case kUp:
+      return blink::PointerData::Change::kUp;
+    case kDown:
+      return blink::PointerData::Change::kDown;
+    case kMove:
+      return blink::PointerData::Change::kMove;
+  }
+  return blink::PointerData::Change::kCancel;
+}
+
+FlutterResult FlutterEngineSendPointerEvent(
+    FlutterEngine engine,
+    const FlutterPointerEvent* pointer) {
+  if (engine == nullptr || pointer == nullptr) {
+    return kInvalidArguments;
+  }
+
+  auto holder = reinterpret_cast<PlatformViewHolder*>(engine);
+
+  blink::PointerData pointer_data;
+  pointer_data.Clear();
+
+  pointer_data.time_stamp = SAFE_ACCESS(pointer, timestamp, 0);
+  pointer_data.change = ToPointerDataChange(
+      SAFE_ACCESS(pointer, phase, FlutterPointerPhase::kCancel));
+  pointer_data.kind = blink::PointerData::DeviceKind::kMouse;
+  pointer_data.physical_x = SAFE_ACCESS(pointer, x, 0.0);
+  pointer_data.physical_y = SAFE_ACCESS(pointer, y, 0.0);
+
+  blink::Threads::UI()->PostTask(
+      [ weak_engine = holder->view()->engine().GetWeakPtr(), pointer_data ] {
+        if (auto engine = weak_engine) {
+          blink::PointerDataPacket packet(1);
```

Comment:
Should we let the client give us an array of pointer data so we can dispatch more than one at a time?